### ISSUE TITLE
logbook: speeding up MVCCScan

### ISSUE THREAD

**Issue by arjunravinarayan** [arjunravinarayan/cockroach#18]
Experiment 1 benchmarked scanning over an SST in two different ways:
first using the regular RocksDB-via-CGo path, and second, using
goleveldb. The result was that goleveldb was faster than
RocksDB-via-CGo. For large kvs both were pretty fast (1GB/s vs 1.5
GB/s), but for small kvs goleveldb was much faster (50MB/s vs
200MB/s).

This shows that there is some juice to be squeezed, but is not
conclusive. Experiment 2 is to set up a more interesting test: to do
the scanning entirely on the C side, and return it in one single call.

### How fast are `WriteBatch`es?

The immediate simple answer is to use the RocksDB `WriteBatch`. But
`WriteBatch` is not the ideal serialization format for a scan-only
workload, so we also want to test to make sure `WriteBatch`
construction is not the bottleneck - so we might also want to just rip
through the scan at the C++ level first to make sure that constructing
the WriteBatch isn't sapping our speed (and if it is, we need to come
up with a better serialization format to return data across the CGo
barrier).

Note that we have already done a basic scan at the C++ level by
calling the `MVCCComputeStats_RocksDB` benchmark, which rips through
the KVs while doing some basic arithmetic (MVCCComputeStats), and that
goes at ~950 MB/s (on a gceworker that has a 250 MB/s SSD, so clearly
being cached).

1. We (with Dan) wrote a little C++ function called `FastScan` that
rips through keys (calling `Next()` until it hits the end key), all
the while putting things into a `WriteBatch` as it scans through.

We changed `BenchmarkFS` to use the exact same data, creating the
`WriteBatch` ran at ~350MB/s. To cross-check our `ComputeStats`
numbers, commenting out the `batch.Put` line put us back at 1200MB/s,
corroborating the `MVCCComputeStats_RocksDB` bench numbers. So
`batch.Put` is a 4x slowdown! We need to investigate whether we can
speed that up, whether we can track pointers into the batch to make
batch construction faster.

But let's hold that thought and start, since that's already faster
than the 50MB/S that we get out of our MVCCIteration today.

### How do we engineer MVCC to use `WriteBatch`es?

We have a few options:

1) Push all the MVCC logic into C++, and return WriteBatches of keys
and values after doing whatever MVCC does (skipping intents, choosing
the correct timestamped KV when there are multiple, etc). This is hard
and lots of work and change is scary and usually bad.

2) Make C++ return WriteBatches, and then rescan over this data in Go,
and retain the current MVCC logic.

This requires 2 passes over the same data (once in C++, and once in
Go), but has us crossing the CGo boundary only once. This is far less
engineering work, and safer, but potentially slower.

3) Make C++ return SSTs, and then rescan over this data in Go and
retain the current MVCC logic.

Now we want to measure the speeds of the 3 options. So we amend
`BenchmarkFS` to also then scan through the `WriteBatch` on the Go
side. This is not perfect, but has us replicating the two-pass
overhead, since we have a pass on the Go side as well. However, we
have to do the work to return the actual `WriteBatch` across the CGo
boundary. In the following results, `goScan` indicates if we scan over
the `WriteBatch` in go after we return it.

```
BenchmarkFS/valueSize=8,_goScan=true-4 1 1045993167 ns/op 64.16 MB/s
BenchmarkFS/valueSize=8,_goScan=false-4 3 375147417 ns/op 178.89 MB/s
BenchmarkFS/valueSize=32,_goScan=true-4 2 630583242 ns/op 106.42 MB/s
BenchmarkFS/valueSize=32,_goScan=false-4 5 280007250 ns/op 239.67 MB/s
BenchmarkFS/valueSize=256,_goScan=true-4 5 237117862 ns/op 283.02 MB/s
BenchmarkFS/valueSize=256,_goScan=false-4 10 165544349 ns/op 405.38
MB/s
```

Option 2 is slower than option 1 (280 vs 400 MB/s) but not terribly
slower. We're going with it. It's still 6x faster.

We briefly tried Option 3, but it was much slower, likely due to the
prefix compression that happens in SSTs by default:

```
BenchmarkFS/valueSize=8-4 2 932186770 ns/op 71.99 MB/s
BenchmarkFS/valueSize=32-4 2 1010337435 ns/op 66.42 MB/s
BenchmarkFS/valueSize=256-4 2 628377277 ns/op 106.80 MB/s
```

We abandoned Option 3 entirely.

# Engineering LogBook:

Now that we decided to go down option 2, its time to engineer this
properly.

`mvcc./go:MVCCScan` ends up calling into `engine/mvcc.go:MVCCIterate`
which seeks the engine to the start point (or the end point in a
reverse scan) and then calls `.Next()` repeatedly while doing its MVCC
logic. We are instead going to call `FastScan()` to get back a
`WriteBatch`, and then create an iterator over that batch and let
`MVCCIterate` work over that iterator. Almost seems too easy. We're
missing something. Now we find out. (Alfonso ominously intones "well
there's a `SimpleIterator` interface and an `Iterator` interface. Are
we sure that `MVCCIterate` doesn't need the latter to be fast, but
`WriteBatch` only implements the former?") It turns out that
`MVCCIterate` requires the `Iterator`, and Alfonso is right, so we
need to implement the remaining functions so that we can iterate
backwards.

Engineering challenges: we need to make `RocksDBBatchReader` a
`SimpleIterator`. This involves plumbing changes to other callers of
`RocksDBBatchReader`.

So once we made `RocksDBBatchReader` a simple iterator and cobbled
together a pipeline, a few additional items came to fore:

1. We need to plumb through "maxKeys" to C, because otherwise we could
get an unbounded read, and then panic on allocating a gigantic slice
to return the WriteBatch. This doesn't happen in practice (because
`replica_command.go` breaks up reads, but it does make testing
difficult.

2. We need to set FastScan to a very simple knob so that we don't have
to constantly recompile.

Apart from this, MVCC iteration via FastScans appear to be slower than
 MVCC iteration via CGo! In order to understand this, I need to set up
 two profiles on a GCEWorker:

A) Profile when not using FastScan B) Profile when using FastScan

The profiles didn't show anything interesting. The two profiles look
similar(???) which is surprising, as FastScan is 4x slower(!!!).

Question: look at the batch sizes returned - is it possible we are
scanning over many more keys than we should be? MVCCScan has a
footguard with a `max` parameter - but `FastScan` does not. So say
`FastScan` returns a batch with 10000 things, but MVCCScan pulls the
plug on MVCCIterate after scanning through 1000 things, and then
resumes the span starting at index 1001. This would result in us
scanning over 10x as many keys as we should be. Check if this is the
case by printing out the batch sizes returned, and summing the total,
and seeing if it matches what we iterate over in the normal
non-FastScan path.

Things deferred for now

1. Reverse scans 2. comprehensive testing 3. passing all the tests (of
note, we fail some tests due to returning incorrect results at the
wrong timestamps)

Things to do later:

1. Does disabling snappy compression on the SST help matters? How do
we turn off Snappy?

2. Why is `WriteBatch.Put()` so slow?

# Picking this back up.

We need to benchmark how fast MVCC scan logic is, since it's a bit of
a black box. If it doesn't go fast enough, we need to speed it
up. Some benchmarking by Alfonso says that it is ~500MB/s (TODO: where
exactly?) but it's time we checked. If we are moving MVCC scan to C++
at some point in the future, we want to understand this anyway.

We start by implementing the `Iterator` interface with a made up
struct that we benchmark can generate and emit keys at lightning
speed.

The problem with this is that our madeup struct needs to implement the
`Iterator` interface efficiently. This is the exact same problem that
`WriteBatch` needs to also do, so we basically need to do this with
the `WriteBatch` interface. So its pretty clear that its our
implementation that can be bottlenecking us.
